SQL: transfer version compatibility decision to the server

bpintea opened the pull request:
This PR transfer the decision to inter-communicate from the clients to the server.
The server will now accept a new request field that conveys the version of the client that originated the request.
For those clients that should have backwards compatibility support (cli, drivers), the server requires that this field be present in query (first/non-paginated) and translate requests.
In this first step, the service policy remains unchanged: the versions need to be in lockstep in order for the server to successfully answer it.

Similarly, the clients still do a server's version validation, but this is just for lower bounding: the server will need to be at least as recent as the client, or newer (plus a concrete lower bound, to prevent connections to older servers missing a version policy).

Closes #52766.

bpintea committed:
add a new version request object parameter

This commit adds a new request object field, "request_version",
containing the version of the requesting client. This parameter is
now accepted by the server and the request is validated against it.
Currently server's and client's versions need to be equal in order for
the request to be accepted.

bpintea committed:
refactor 'clientVersion' to 'version'; more tests

- change 'clientVersion' to 'version' in function names and members;
- add testing for the validation of client's version.

bpintea committed:
propagate the 'client_version' to more tests

Add the new 'client_version' field into the tests.

Also:
- add the 'client_version' field into proto/SqlClearCursorRequest.java
builder.
- remove one unused member (ConnectionConfiguration.java:CLIENT_ID).

bpintea committed:
update version checking. update tests

This commit updates the clients to only check server's version against a
lower bound: the server needs to always be on a higher version than the
client.

The server now enforces the version backwards compatibility. Currently,
the versions still need to be in lockstep. Laxing this would be part of
future work.

The tests are updated to reflect the policy change and to have the
version always included in query requests.

bpintea committed:
further tests fixing

- add the version field into the request object of more tests

elasticmachine commented:
Pinging @elastic/es-search (:Search/SQL)

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/add_version_request_field

astefan reviewed (commented):
Left a first set of comments.

bpintea committed:
update renamed method assertThrows

s/assertThrows/assertRequestBuilderThrows as per #52582

bpintea commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Version.java:
```diff
@@ -0,0 +1,134 @@
+
+
+
+    public static final Version V_7_7_0 = new Version(7, 7, 0);
```
It could make sense, yes, but I imagine that we'll need to log a new version here with each "breaking" change we add in the protocol. If we'd instrument somehow these changes, referring to the versions by their number might be easier. But we could also refer to the versions by a name that conveys the gist of the breaking change, true. Wdyt?

costin replied:
While I would prefer a name, I think using the versions gives us excellent flexibility in the long term. I'd opt towards adding a comment plus an utility method (that uses the field internally) with a verbose name:
```
public boolean supportsVersioning(Version version) {
  return version.compareTo(V_7_...) >= 0;
}
```

bpintea commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Mode.java:
```diff
@@ -33,4 +33,9 @@ public String toString() {
     }
+
+    // TODO: replace all "== Mode.CLI"?
+    public static boolean isCli(Mode mode) {
```
In many cases the code look like `Mode.isDriver(mode) || mode == Mode.CLI` and `Mode.isDriver(mode) || Mode.isCli(mode)` looked more consistent to me. I got to it from a first thought of adding a `Mode.isStrictClient()` to check for all drivers+cli (since we treat those grouped in many cases). Irrespective of that, though, if Mode being an enum is to be considered, one would either stick to `mode == Mode.JDBC || mode == Mode.jdbc || mode == Mode.CLI` _or_ the two methods wrapping, IMO.
However, this isn't a strictly necessary refactoring, it is subjective and I'll happily change it back if you still think the old way is better (just 👍 this comment).

bpintea commented on x-pack/plugin/sql/sql-cli/src/test/java/org/elasticsearch/xpack/sql/cli/CliSessionTests.java:
```diff
@@ -44,6 +44,31 @@ public void testConnection() throws Exception {
+                Version.fromString(major + "." + minor + ".23").toString(),
+                ClusterName.DEFAULT.value(), UUIDs.randomBase64UUID()));
+        CliSession cliSession = new CliSession(httpClient);
+        expectThrows(ClientException.class, cliSession::checkConnection);
```
Sure. Neither of the already defined tests check the message, but I'll look into adding it.

bpintea replied:
added.

bpintea commented on x-pack/plugin/sql/sql-cli/src/test/java/org/elasticsearch/xpack/sql/cli/CliSessionTests.java:
```diff
@@ -26,7 +26,7 @@
     public void testProperConnection() throws Exception {
         HttpClient httpClient = mock(HttpClient.class);
-        when(httpClient.serverInfo()).thenReturn(new MainResponse(randomAlphaOfLength(5), org.elasticsearch.Version.CURRENT.toString(),
+        when(httpClient.serverInfo()).thenReturn(new MainResponse(randomAlphaOfLength(5), Version.CURRENT.toString(),
```
The class has been making use of the `client.Version` class already. Using two Version versions is not strictly needed anymore. Before this PR `client.Version.toString()` behaved differently than `elasticsearch.Version.toString()`, but this has been changed (old behavior is still available in `client.Version.versionString()`, but that's not used anywhere atm.)

matriv commented on x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/RestSqlSecurityIT.java:
```diff
@@ -233,15 +237,16 @@ public void testHijackScrollFails() throws Exception {
         final String m = randomMode();
         ResponseException e = expectThrows(ResponseException.class, () -> RestActions.runSql("full_access",
-                new StringEntity("{\"cursor\":\"" + cursor + "\"" + mode(m) + "}", ContentType.APPLICATION_JSON), m));
+                new StringEntity("{\"cursor\":\"" + cursor + "\"" + mode(m) + version(mode) + "}", ContentType.APPLICATION_JSON),
```
should it be `version(m)` here?

bpintea replied:
nice catch, thanks. fixed.

matriv commented on x-pack/plugin/sql/qa/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/RestSqlIT.java:
```diff
@@ -42,16 +42,16 @@ public void testErrorMessageForTranslatingSQLCommandStatement() throws IOExcepti
         expectBadRequest(() -> runTranslateSql(
-            "{\"query\":\"SELECT null WHERE 0 = ? \", \"mode\": \"odbc\", \"params\":[{\"type\":\"invalid\", \"value\":\"irrelevant\"}]}"
-            ),
+            "{\"query\":\"SELECT null WHERE 0 = ? \", \"mode\": \"odbc\" " + version("odbc") +
```
maybe `mode("odbc")` as well (also below)?

bpintea replied:
makes sense, fixed.

matriv commented on x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlRequest.java:
```diff
@@ -56,8 +58,9 @@ public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
         out.writeEnum(requestInfo.mode());
         out.writeOptionalString(requestInfo.clientId());
+        out.writeOptionalString(requestInfo.version() == null ? null : requestInfo.version().toString());
```
since you use `writeOptionalString` no need for null check.

bpintea replied:
the check is there for the `.toString()` to be safe.

matriv replied:
ah yep, I see, thx!

matriv commented on x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlRequestParsersTests.java:
```diff
@@ -66,10 +68,11 @@ public void testClearCursorRequestParser() throws IOException {
+        request = generateRequest("{\"cursor\" : \"whatever\", \"client_id\" : \"CLI\", \"client_version\": \"1.2.3\"}",
                 SqlClearCursorRequest::fromXContent);
         assertNull(request.clientId());
+        assertNull(request.version());
```
Why is it null here?

bpintea replied:
b/c the parser for `SqlClearCursorRequest` doesn't set these fields.

matriv commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Version.java:
```diff
@@ -0,0 +1,134 @@
+    public static final int MINOR_MULTIPLIER = REVISION_MULTIPLIER * REVISION_MULTIPLIER;
+    public static final int MAJOR_MULTIPLIER = REVISION_MULTIPLIER * MINOR_MULTIPLIER;
+
+
```
nit: please keep one or two empy lines.

bpintea replied:
sure. (i'm trying to to commit whitespace editor "fixes" - as everybody, prolly - and this one slipped through.)

matriv reviewed (commented):
Nice stuff! Left some comments.

bpintea committed:
address review notes

- add exception message testing;
- remove superfluous empty lines;
- fix redefinition of mode string;

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/add_version_request_field

costin commented on x-pack/plugin/sql/qa/multi-node/src/test/java/org/elasticsearch/xpack/sql/qa/multi_node/RestSqlMultinodeIT.java:
```diff
@@ -110,7 +111,7 @@ private void assertCount(RestClient client, int count) throws IOException {
 
         Request request = new Request("POST", SQL_QUERY_REST_ENDPOINT);
-        request.setJsonEntity("{\"query\": \"SELECT COUNT(*) FROM test\"" + mode(mode) + "}");
+        request.setJsonEntity("{\"query\": \"SELECT COUNT(*) FROM test\"" + mode(mode) + version(mode) + "}");
```
It's worth adding an utility (or two) to handle the query wrapping:
1. one to add mode() and version at the end as parameters
2. another to wrap the given string in a query with the `{\"query\"`, escaped quotes and everything else.

bpintea replied:
That's true. And the fact that the queries are "manually" build can lead to more work whenever the fields list is updated. I've had a stab at mimicking SqlQueryRequestBuilder's interface at building requests and apply the change through the tests [here](https://github.com/bpintea/elasticsearch/commit/392dfc275d00dc420c2907e2405f88cbfeac4b12), but the change is a bit larger, and it might be worth having a separate PR?

costin commented on x-pack/plugin/sql/qa/security/src/test/java/org/elasticsearch/xpack/sql/qa/security/RestSqlSecurityIT.java:
```diff
@@ -70,10 +71,10 @@ public void expectMatchesAdmin(String adminSql, String user, String userSql) thr
             String mode = randomMode();
             Map<String, Object> adminResponse = runSql(null,
-                    new StringEntity("{\"query\": \"" + adminSql + "\", \"fetch_size\": 1" + mode(mode) + "}",
+                    new StringEntity("{\"query\": \"" + adminSql + "\", \"fetch_size\": 1" + mode(mode) + version(mode) + "}",
```
Same comment as above, it applies in other places but I won't repeat it.

costin commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/SqlProtocolTestCase.java:
```diff
@@ -252,14 +253,13 @@ private void assertQuery(String sql, String columnName, String columnType, Objec
             requestContent = new StringBuilder(requestContent)
                     .insert(requestContent.length() - 1, ",\"binary_format\":" + binaryCommunication).toString();
-            binaryCommunication = ((Mode.isDriver(m) || m == Mode.CLI) && binaryCommunication);
+            binaryCommunication = ((Mode.isDriver(mode) || Mode.isCli(mode)) && binaryCommunication);
```
`Mode.isCli` doesn't add much value and since `Mode.isDriver() || Mode.isCli` is a common pattern, it's worth wrapping that into its own method: `Mode.isInternalClient` or `Mode.isDedicatedClient` or `Mode.isSqlClient` to differentiate between our "own" clients and everything else.

bpintea replied:
Agreed, fixed it.

costin commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlUsageTestCase.java:
```diff
@@ -255,18 +256,18 @@ private void runTranslate(String sql) throws IOException {
         }
 
-        runSql(mode, clientType, sql);
+        runSql(mode.toString(), clientType, sql);
```
👍

costin commented on x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/AbstractSqlQueryRequest.java:
```diff
@@ -57,6 +63,7 @@
     static final ParseField FILTER = new ParseField("filter");
     static final ParseField MODE = new ParseField("mode");
     static final ParseField CLIENT_ID = new ParseField("client_id");
+    static final ParseField CLIENT_VERSION = new ParseField("client_version");
```
I would just say `version` since all parameters are client specific. `client_id` has is as such since just `id` would have been too cryptic.
There's no client_mode, or client_params so why client_verision.

In fact, if anything we could change client_id to drop the prefix by using a different word: something like fingerprint or info (too loose).

bpintea replied:
Makes sense, fixed it.

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/RequestInfo.java:
```diff
@@ -35,14 +35,25 @@
+        version(version);
+    }
+
+    public RequestInfo(Mode mode, Version version) {
```
It seems this constructor is used everywhere using `Version.CURRENT` - maybe it makes sense to automatically initialize version when Mode is passed instead of setting it as `null`.
As it stands the constructor initializes Version twice.

bpintea replied:
> As it stands the constructor initializes Version twice.

Addressed this.
But auto-initialising along with Mode is not possible, since the `CURRENT` Version value is only available in `client`, but not in `proto`.

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/SqlQueryRequest.java:
```diff
@@ -180,6 +180,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
             builder.field("client_id", clientId());
         }
+        if (version() != null) {
+            builder.field("client_version", version().toString());
```
Some constants for these strings would be nice.

bpintea replied:
Very much agreed. I would address this in a follow-up PR, thought, since these hardcoded strings are used extensively in tests?

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/Version.java:
```diff
@@ -16,37 +16,20 @@
 import java.util.jar.Manifest;
 
-public class Version {
+public class Version extends org.elasticsearch.xpack.sql.proto.Version {
```
Functionally, this class is not another `Version` but a consumer of `Version`.
It has only `isServerCompatible` and `jdbcMajor/Minor` (which were added to this class since it was the only one but could be moved somewhere else).

Try refactoring it to something else either a consumer:

`ClientVersion(Version) { .. }`
or if there's no state, utility methods - ClientVersion/VersionUtils.isServerCompatible(version)`

bpintea replied:
Sure. I've turned that class into a `ClientVersion` utility class.

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/Version.java:
```diff
@@ -117,11 +100,21 @@ static Version extractVersion(URL url) {
+    // as well.
+    public static boolean isServerCompatible(Version server) {
+        // Starting with this version, the compatibility logic moved from the client to the server. Only relevant for 7.x releases.
+        return server.compareTo(Version.V_7_7_0) >= 0
```
The V_7_7_0 constant is leaking - encapsulate that into Version (which already has the constant) to something like `Version.hasVersionCompatibility/isCompatible` or something along those lines so the lines becomes:
`return server.isCompatible()`

bpintea replied:
took on your `.hasVersionCompatibility()` suggestion.

costin commented on x-pack/plugin/sql/sql-client/src/test/java/org/elasticsearch/xpack/sql/client/HttpClientRequestTests.java:
```diff
@@ -158,7 +158,7 @@ private void assertBinaryRequestForDrivers(boolean isBinary, XContentType xConte
                 randomBoolean(),
                 randomAlphaOfLength(128),
-                new RequestInfo(mode),
+                new RequestInfo(mode, Version.CURRENT),
```
See my comment on `RequestInfo` constructor.

costin reviewed (commented):
Left a first round of comments.

bpintea committed:
address review comments

- change client.Version to a utility client.ClientVersion class;
- rename proto.Version to proto.SqlVersion;
- remove double initialization of 'version' in RequestInfo;
- make V_7_7_0 version private and expose methods to compare versions
  against that;
- add a Mode.isDedicatedClient(), remove Mode.isCli();
- renamed "client_version" field to just "version".

bpintea committed:
erge branch 'feat/add_version_request_field' of github.com:bpintea/elasticsearch into feat/add_version_request_field

bpintea committed:
add change skipped in last commit

- VersionTests also updated with new ClientVersion class.

bpintea commented:
@elasticmachine run elasticsearch-ci/2

bpintea committed:
adapt u-test to relaxed client-side version checks

The client now only makes sure that server's version is above a certain
limit (and not that that version is lower than own).

bpintea committed:
Merge remote-tracking branch 'origin/master' into feat/add_version_request_field

astefan commented on x-pack/plugin/sql/sql-action/src/main/java/org/elasticsearch/xpack/sql/action/SqlQueryRequest.java:
```diff
@@ -72,7 +72,7 @@ public SqlQueryRequest(String query, List<SqlTypedParamValue> params, QueryBuild
     @Override
     public ActionRequestValidationException validate() {
-        ActionRequestValidationException validationException = null;
+        ActionRequestValidationException validationException = super.validate();
```
Why wasn't this called here before? The code we had was a bug?

bpintea replied:
The overridden function (in `action/AbstractSqlQueryRequest.java`) wasn't implemented before (and now it checks client's version).

astefan commented on x-pack/plugin/sql/sql-action/src/test/java/org/elasticsearch/xpack/sql/action/SqlRequestParsersTests.java:
```diff
@@ -119,12 +127,16 @@ public void testQueryRequestParser() throws IOException {
                 + "\"request_timeout\":\"5s\",\"page_timeout\":\"10s\"}", SqlQueryRequest::fromXContent);
         assertNull(request.clientId());
         assertEquals(randomMode, request.mode());
+        if (clientVersion.isEmpty() == false) {
```
Why not using the same condition as above? `Mode.isDedicatedClient(randomMode)`

bpintea replied:
good point.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/RestSqlQueryAction.java:
```diff
@@ -60,7 +60,7 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
         String accept = null;
 
-        if ((Mode.isDriver(sqlRequest.requestInfo().mode()) || sqlRequest.requestInfo().mode() == Mode.CLI)
+        if ((Mode.isDedicatedClient(sqlRequest.requestInfo().mode()))
```
Too many round brackets?

bpintea replied:
thanks, fixed.

astefan reviewed (commented):
Left few more comments.

bpintea committed:
address review comments

- remove extra brackets
- replace condition from string checking to client type checking

bpintea commented:
@elasticmachine run elasticsearch-ci/2

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -16,45 +18,17 @@
-    public static Version fromString(String version) {
-        return new Version(version, "Unknown", from(version));
-    }
+public class ClientVersion {
```
Worth adding a javadoc to clarify the difference between `ClientVersion`, `SqlVersion` and `Version`.

matriv replied:
+1

bpintea replied:
I've added a short description for `SqlVersion` and `ClientVersion`. Hope it's enough, but could elaborate if felt as necessary.

costin reviewed (approved):
LGTM

bpintea committed:
add javadoc on the *Version classes

Add short description of the SqlVersion and ClientVersion classes.

bpintea committed:
fix javadoc reference

fix ref to unavailable org.elasticsearch.Version class

matriv reviewed (approved):
LGTM

astefan commented on x-pack/plugin/sql/sql-cli/src/test/java/org/elasticsearch/xpack/sql/cli/CliSessionTests.java:
```diff
@@ -26,40 +27,58 @@
         when(httpClient.serverInfo()).thenThrow(new SQLException("Cannot connect"));
         CliSession cliSession = new CliSession(httpClient);
-        expectThrows(ClientException.class, cliSession::checkConnection);
+        expectThrows(ClientException.class, cliSession::checkConnection, "java.sql.SQLException: Cannot connect");
```
That's a good way of wrapping two checks in the same one, but throughout the code in SQL (where there are many such checks) we do them separately: `expectThrows` and then another `assertEquals` on the exception message.

I'm not saying you should change it, but the way it is now, it is inconsistent throughout sql code.

bpintea replied:
adjusted to keep consistency.

astefan commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java:
```diff
@@ -16,45 +18,25 @@
+/**
+ * Clients-specific version utility class.
+ * <p>
+ *     The class provides the SQL clients the version identifying the release their are part of. The version is read from the encompassing
```
`they're` or `they are` in the sentence `identifying the release their are part of`.

bpintea replied:
thanks, fixed.

astefan commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/Mode.java:
```diff
@@ -33,4 +33,8 @@ public String toString() {
     }
+
+    public static boolean isDedicatedClient(Mode mode) {
+        return mode == JDBC || mode == ODBC || mode == CLI;
```
`return isDriver(mode) || mode == CLI`

astefan reviewed (approved):
Left some minor comments still, but LGTM.

bpintea committed:
adress review comments

- fix spelling
- remove two-liner function
